Verify using sniffed packets instead of logs in link metric query and client capability query tests

aciliketcap opened the pull request.

aciliketcap commented:
This PR is part of #1308, it utilizes sniffer in two more tests.

aciliketcap commented:
Do we expect tests to fail at first issue (like most unit tests) or should they run until they can report each and every issue they encountered? Is it OK if they throw unhandled exceptions or should they mark themselves as failed and then log and swallow any expected exceptions?

I wrote the tests reporting as much issues as they can and never raising exceptions (that we expect).

VladyslavTupikin commented:
Hi. @aciliketcap  Can you please prepare commits according to the instruction in CONTRIBUTING.md?

Commit:
tests: wrap exceptions thrown by individual tests

This way script can continue executing the next test even if one of
tests throws an exception. Also we can use assert statements inside
tests since AssertionErrors are handled and they fail the test with an
assertion failed message.

If we want to fail the test when an assertion has failed we can do the
following:

assert False, "Some assertion failed and we don't want to continue"
produces
FAIL: Some assertion failed and we don't want to continue

If we don't add an assertion failed message a generic "Assertion failed"
message will be printed, so please add messages to assert statements.

assert False
produces
FAIL: Assertion failed

But, if self.fail() is called somewhere with a proper message and we
want to end the test there without adding anything to that message you
can assert without a message.

self.fail("Something failed and we don't want to go on")
assert False
produces
FAIL: Something failed and we don't want to go on

When a test throws an unexpected run-time error the test is marked as
failed and a stack trace is printed in debug mode.

FAIL: Test failed unexpectedly: SomeRuntimeException('some error
message')
Traceback (most recent call last):
  ... stack trace showing the method
  ... which has thrown the unhandled exception
SomeRuntimeException: some error message

Please make sure to wrap all exceptions and throw only AssertionErrors
to fail the tests.

Signed-off-by: Sinan Akpolat <sinan.akpolat@lifemote.com>

abelog commented on tests/test_flows.py:
```diff
@@ -814,53 +834,131 @@ def test_client_capability_query(self):
-        env.controller.dev_send_1905(env.agents[0].mac, 0x8009,
-                                     tlv(0x90, 0x000C,
-                                         '{} {}'.format(env.agents[0].radios[0].mac, sta1.mac)))
+        mid1 = env.controller.dev_send_1905(env.agents[0].mac, 0x8009,
```
*Nitpick*: i really don't like such senseless variables.
```suggestion
        mid_cap_query = env.controller.dev_send_1905(env.agents[0].mac, 0x8009,
```
Will have more sense and easier to read, IMHO.

aciliketcap replied:
I kept it as mid since the parameter name in check_cmdu_type method is mid and since it is not really used in test code.

abelog commented on tests/test_flows.py:
```diff
@@ -814,53 +834,131 @@ def test_client_capability_query(self):
-        self.check_log(env.controller,
-                       r"Result Code= FAILURE, client MAC= {}, BSSID= {}"
-                       .format(sta1.mac, env.agents[0].radios[0].mac))
+        query1 = self.check_cmdu_type_single("client capability query", 0x8009, env.controller.mac,
```
Same nitpick about sensless naming of variable.
```suggestion
        client_cap_query = self.check_cmdu_type_single("client capability query", 0x8009, env.controller.mac,
```

aciliketcap replied:
yeah, I guess I can make this more meaningful, like query_for_error instead of query1 and query_for_success instead of query2.

arnout replied:
I think "query" would be OK, but as soon as you need more than one, "query1" and "query2" are meaningless.

In this specific case, however, I think it would be better to split the test into two separate tests. The two parts of the test are pretty much independent of each other.

aciliketcap replied:
I changed the variable names in that test. This time the lines are too long but I think it is OK for test code, what we are testing should be clear in variable names.

Btw I didn't used client_cap_query name since it is evident in the test name and check_cmdu_type error string parameter. I instead changed the name to reflect that we are testing for a query which returns a report with error and then a query with a successful report.

abelog commented on tests/test_flows.py:
```diff
@@ -814,53 +834,131 @@ def test_client_capability_query(self):
+                    if query2_tlv.client_info_bssid != env.agents[0].radios[0].mac:
+                        self.fail("Wrong bssid in query")
+                except AttributeError:
+                    self.fail("Query does not contain bssid")
```
This block repeats frequently, should it be a separate method? It will significantly improve readability.

aciliketcap replied:
All of them are removed since I'm using assert statements now.

abelog commented on tests/test_flows.py:
```diff
@@ -814,53 +834,131 @@ def test_client_capability_query(self):
+                        self.fail("Capability report result is not successful")
+                    if report2_cap_report_tlv.client_capability_frame != association_frame:
+                        self.fail("Capability report does not contain correct frame")
+                        debug(f"Frame received\n{report2_cap_report_tlv.client_capability_frame}")
```
*Nitpick*: mainly, we tend to use `format` for string formatting.

aciliketcap replied:
Oh, if most of the code is written that way I can change it too. But did you consider using format strings instead of .format() ? Many people (for example devs in my  current company :)) like format strings more.

arnout replied:
Unfortunately, Intel is still using build machines with Python 3.4.

aciliketcap replied:
Then I'll definitely remove all format strings.

abelog reviewed (changes requested):
I still think that we should avoid repeat of code blocks. So not approving yet.
*Nitpicks* are optional to fix.

arnout commented on tests/test_flows.py:
```diff
@@ -695,17 +695,37 @@ def test_ap_capability_query(self):
+        if resp:
+            debug("Checking response contains both Tx and Rx metrics")
+            self.check_cmdu_has_tlvs(resp, 0x09)
+            self.check_cmdu_has_tlvs(resp, 0x0a)
```
We can check a little more about these TLVs. E.g. at least the controller should be a neighbour, so it should be mentioned in both metrics.

aciliketcap replied:
I will be checking both links to controller and the other agent using what we know in env.

arnout commented on tests/test_flows.py:
```diff
@@ -834,53 +834,131 @@ def test_client_capability_query(self):
-e1 09 00 bf 0c b0 79 d1 33 fa ff 0c 03 fa ff 0c
-03 c7 01 10 """
-
+        association_frame = "00:0e:4d:75:6c:74:69:2d:41:50:2d:32:34:47:2d:31:"\
```
Since you're changing this anyway, it would make sense to also move it down to the place where you actually need it (i.e. where the station associates).

aciliketcap commented:
The CI flow gave flake8 error but it does not appear when I run flake8 with 'flake8 --config .flake8 tests/' ?

arnout commented on tests/test_flows.py:
```diff
@@ -834,53 +834,131 @@ def test_client_capability_query(self):
+        report1 = self.check_cmdu_type_single("client capability report", 0x800a,
+                                              env.agents[0].mac, env.controller.mac, mid1)
+        if report1:
+            report1_client_info_tlv = self.check_cmdu_has_tlvs(report1, 0x90)[0]
```
I don't like that you take `[0]` here - especially since the order of the TLVs is not defined. So either there is just one, and you should use check_cmdu_has_tlv_single, or there are several and you should iterate over them to find the one you want to check.

In this case, since we only requested a single station, there must be only a single client info tlv, so _single is appropriate.

Same issue occurs in a few other places.

arnout commented on tests/test_flows.py:
```diff
@@ -278,6 +279,10 @@ def run_tests(self, tests):
+            except AssertionError as ae:
+                self.fail(f"{ae}")
+            except Exception as e:
+                self.fail(f"Test has thrown an exception\n{traceback.format_exc()}")
```
I don't think we should print the full traceback. Also, I think it's worth repeating the test name.
```suggestion
                self.fail("Test {} has thrown an exception\n{}".format(test, e))
```

(Incidentally, this will also fix the flake8 error because `e` is now used)

With that, there's also no real need to handle the AssertionError differently.

BTW, it would be nice to show in the commit message some examples of what the output will be when an exception is handled. Something like:
```
For example, when a non-existing TLV field is accessed, the following failure will be printed:
...
```
For reviewers, that's much easier to help decide if it is done properly.

aciliketcap replied:
Ah, I totally missed that this was causing the flake8 error.

aciliketcap replied:
I expect tests to throw AssertionError when they fail and decide to quit. However I consider other exceptions as unexpected run-time exceptions and therefore I conclude that the test code has failed, not the DUT. I think we definitely need the stack trace (in debug / verbose mode) so that we can have a chance to figure out what went wrong without having to debug the test.

arnout commented on tests/test_flows.py:
```diff
@@ -707,30 +707,28 @@ def test_link_metric_query(self):
-        if query:
-            query_tlv = self.check_cmdu_has_tlv_single(query, 0x08)
+        query_tlv = self.check_cmdu_has_tlv_single(query, 0x08)
+        assert query_tlv
```
I would actually move the assert inside the check_ function. Yes, there are some cases where we actually could continue the test even if the CMDU is missing, but that's not the common case. In the few cases where we do want to continue, we can still add a try block.

And actually, the assertion isn't even needed, because it's implied when you start using query_tlv below.

arnout commented on tests/test_flows.py:
```diff
@@ -707,30 +707,28 @@ def test_link_metric_query(self):
-            except (AttributeError, ValueError):
-                self.fail("Requested metrics are not recognized")
+        debug("Checking query type and queried metrics are correct")
+        assert query_tlv.link_metric_query_type, "Query type is not recognized"
```
This is not needed. We can pretty much assume that it's there, otherwise the tlv wouldn't exist to begin with. And if it doesn't, the following statement will simply raise an AttributeError.

aciliketcap replied:
I do that to prevent the AttributeError, test should only fail with AssertionError I think, otherwise we can't discern if the test failed or the test code failed.

arnout replied:
It doesn't actually prevent the attribute error - the assert will never even trigger if the attribute doesn't exist. You'd need
```
assert hasattr(query_tlv, "link_metric_query_type")
```

However, I think it is much more convenient if we can just access attributes without explicit asserts. In this case, we could actually implement a `__getattr__` function on TlvStruct that asserts.

arnout commented on tests/test_flows.py:
```diff
@@ -707,30 +707,28 @@ def test_link_metric_query(self):
+        debug("Checking query type and queried metrics are correct")
+        assert query_tlv.link_metric_query_type, "Query type is not recognized"
+        assert int(query_tlv.link_metric_query_type) == 0x00, \
+            "Query type is not 'All neighbors'"
```
I don't think we should use asserts everywhere. It's better if we can continue the test, so assert isn't always good. And in this case I actually think the original code is easier to read.

arnout commented on tests/test_flows.py:
```diff
@@ -852,31 +852,33 @@ def test_client_capability_query(self):
-        mid1 = env.controller.dev_send_1905(env.agents[0].mac, 0x8009,
-                                            tlv(0x90, 0x000C, '{} {}'.format(
-                                                env.agents[0].radios[0].mac, sta1.mac)))
+        mid_for_error = env.controller.dev_send_1905(env.agents[0].mac, 0x8009,
```
As noted elsewhere, this is not needed if we simply split the test into two tests.

Also, changes that are a result of review should typically be done as fixups. So in this case, do an interactive rebase and edit the commit that introduces mid1 and fix it there.

arnout reviewed (dismissed):
Pretty good PR for your first contribution!

Myself, I would have put the commit that adds the exception handlers first, so the "use sniffed data" commits could immediately use the new approach. However, I kind of like this style as well, because it clearly shows why such a refactoring is needed - it's a bit of the [OneTwoThreeRefactor](http://c2.com/cgi/wiki?OneTwoThreeRefactor) approach.

Our review workflow is that I already approve now, even though there are still plenty of review comments to address. By approving already, you can simply make the requested changes and merge right away. And if there is something you're not sure of or disagree with, we discuss it either in the comments or in Slack before merging.

abelog commented:
Removing requested changes mark as branch is rebuild by author.

arnout commented on tests/test_flows.py:
```diff
@@ -278,6 +279,18 @@ def run_tests(self, tests):
+                    if not str(ae):
+                        pass
+                else:
+                    self.fail("{}".format(ae))
```
The True branch of the condition does nothing at all, so
```suggestion
                if not self.check_error:
                    self.fail("{}".format(ae))
```

However, I don't see why you would depend on check_error. If the assertion came after a completely unrelated failure, you still want to mark the failure. So I'd just do:

```suggestion
                if str(ae):
                    self.fail("{}".format(ae))
```

aciliketcap replied:
I still wanted test to fail if self.fail() is not called before (inside test method) and there is no assertion failure message. I ended up printing a generic "Assertion failed" message in that case.

Commit:
tests: raise exception instead of returning None in check CMDU methods

Signed-off-by: Sinan Akpolat <sinan.akpolat@lifemote.com>

Commit:
tests: test_link_metric_query: use sniffer and assert statements

Signed-off-by: Sinan Akpolat <sinan.akpolat@lifemote.com>

Commit:
tests: test_client_capability_query: split the test and use sniffer and assert statements

The test is split into two tests. One of them is testing the case
where there is no client connected and client capability query is
replied with a report containing an error. The other one is testing the
case where one client is connected and its association frame is
verified.

The base of the test where client capability query is sent and a valid
report is returned is tested in base_test_capability_query to avoid code
duplication.

Signed-off-by: Sinan Akpolat <sinan.akpolat@lifemote.com>

Commit:
tests: add method to check fail conditions without terminating the test

Prepare for "tests: test_link_metric_query: use assert statements" and
"tests: add method to check fail conditions without terminating the
test"

Signed-off-by: Sinan Akpolat <sinan.akpolat@lifemote.com>

abelog commented on tests/test_flows.py:
```diff
@@ -266,6 +269,13 @@ def check_topology_notification(self, eth_src: str, neighbors: list,
 
         return True
 
+    def safe_check_obj_attribute(self, obj, attrib_name, expected_val, fail_str):
```
*Nitpick*: please document arguments types. Makes usage of this method easier for others.

abelog commented on tests/test_flows.py:
```diff
@@ -114,20 +116,19 @@ def check_cmdu_type(
         cmdus = self.check_cmdu_type(msg, msg_type, eth_src, eth_dst, mid)
         if not cmdus:
-            return None  # Failure already reported by check_cmdu
+            assert False  # Failure already reported by check_cmdu
```
*Nitpick*: We tend to have commentaries above the string, not inline.

arnout replied:
No, we have both actually.

abelog commented on tests/test_flows.py:
```diff
@@ -810,58 +846,83 @@ def test_combined_infra_metrics(self):
-    def test_client_capability_query(self):
-        sta1 = env.Station.create()
-        sta2 = env.Station.create()
+    def base_test_client_capability_query(self, sta):
```
*Nitpick*: please document arguments types. Makes usage of this method easier for others.

abelog commented on tests/test_flows.py:
```diff
@@ -810,58 +846,83 @@ def test_combined_infra_metrics(self):
-                       .format(sta1.mac, env.agents[0].radios[0].mac))
+        cap_report_tlv = self.check_cmdu_has_tlv_single(report, 0x91)
+        self.safe_check_obj_attribute(cap_report_tlv, 'client_capability_result', '0x00000001',
+                                      "Capability report does not contain error")
```
To make message more informative.
```suggestion
                                      "Capability query was successful for disconnected STA")
```

abelog commented on tests/test_flows.py:
```diff
@@ -810,58 +846,83 @@ def test_combined_infra_metrics(self):
+                                          "Capability report result is not successful")
+            try:
+                if cap_report_tlv.client_capability_frame != expected_association_frame:
+                    self.fail("Capability report does not contain correct frame")
```
Just small suggestion.
```suggestion
                    self.fail("Capability report does not contain expected frame")
```

abelog commented on tests/test_flows.py:
```diff
@@ -810,58 +846,83 @@ def test_combined_infra_metrics(self):
+            try:
+                if cap_report_tlv.client_capability_frame != expected_association_frame:
+                    self.fail("Capability report does not contain correct frame")
+                    debug(f"Frame received\n{cap_report_tlv.client_capability_frame}")
+                    debug(f"Frame expected\n{expected_association_frame}")
```
Seems, those debug messages should be before `self.fail` call.

aciliketcap replied:
Yeah, but I though when you are reading you'd first want to be informed that the test has failed and then how it failed. If self.fail() made the test fail then of course I'd have to do it the other way around.

abelog reviewed (approved):
Nitpicks only, nothing crucial. In general - LGTM.

arnout commented on tests/test_flows.py:
```diff
@@ -80,6 +80,7 @@ def check_cmdu(
         result = env.wired_sniffer.get_cmdu_capture(match_function)
         if not result:
             self.fail("No CMDU {} found".format(msg))
+            assert False
```
(nitpick)
```suggestion
        assert result, "No CMDU {} found".format(msg)
```

You already had something like this below, so I guess it was just an oversight :-) But there are still other cases that could be simplified like this.

arnout commented on tests/test_flows.py:
```diff
@@ -174,24 +175,25 @@ def check_cmdu_has_tlvs(
-            return None
+            assert False
         if len(tlvs) > 1:
             self.fail("More than one ({}) TLVs of type 0x{:02x} found".format(len(tlvs), tlv_type))
             debug("  {}".format(packet))
-            return None
+            assert False
```
Hm, it would be nice if we had something like `ifdebug` that allows us to do:
```
assert len(tlvs) == 1, "More than one ({}) TLVs of type 0x{:02x} found{}".format(
    len(tlvs), tlv_type, ifdebug("\n  " + str(packet)))
```

Not for this PR, probably.

aciliketcap replied:
We can prepare a string before hand (empty string if not debug) and then print it. Doing this in a separate function as you suggested is cleaner of course.

arnout commented on tests/test_flows.py:
```diff
@@ -174,24 +175,25 @@ def check_cmdu_has_tlvs(
     def check_cmdu_has_tlvs_exact(
         self, packet: Union[sniffer.Packet, None], tlvs: [sniffer.Tlv]
-    ) -> None:
+    ) -> bool:
```
If you change this in asserts as well, there's no need to correct the return type.

aciliketcap replied:
I did it. This method is not used btw :shrug:

arnout commented on tests/test_flows.py:
```diff
@@ -716,17 +716,32 @@ def test_ap_capability_query(self):
 
-        debug("Confirming link metric query has been received on agent")
-        self.check_log(env.agents[0], "Received LINK_METRIC_QUERY_MESSAGE")
+        query = self.check_cmdu_type_single("link metric query", 0x0005,
```
Commit message is not correct... It says "use assert statements" but you're actually using the sniffer instead of check_log.

aciliketcap replied:
You are absolutely right

arnout commented on tests/test_flows.py:
```diff
@@ -846,58 +846,83 @@ def test_combined_infra_metrics(self):
+            self.safe_check_obj_attribute(cap_report_tlv, 'client_capability_result', '0x00000000',
+                                          "Capability report result is not successful")
+            try:
+                if cap_report_tlv.client_capability_frame != expected_association_frame:
```
Doesn't safe_check_obj_attribute work here? Or is it just for the debug info that you have to special case it?

aciliketcap replied:
Yes, that's because I wanted to stop and print debug info if attribute exists but it is wrong. Maybe I can return False when safe_check_attribute fails and write it like;
if safe_check_obj_attrib():
  if cap_report.frame != expected_frame:
    debug_stuff()

arnout reviewed (approved):
*Excellent* commit messages! And in general, this is a pretty good pull request.

In the future though I think it's better to merge things a little sooner and improve in follow-up pull requests. Review feedback doesn't necessarily mean that you have to make those changes right away.
